compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUEnums.td

```diff
@@ -98,15 +98,19 @@ class IREEGPU_I32MmaEnumAttr<string name, string summary, list<I32EnumAttrCase>
   let genSpecializedAttr = 0;
 }
 
+// Format: <kind>_<input-type>_<M>x<N>x<K>_<output-type>
 def MFMA_F16_16x16x16_F32 : I32EnumAttrCase<"MFMA_F16_16x16x16_F32", 0>;
-def MFMA_F16_32x32x8_F32 : I32EnumAttrCase<"MFMA_F16_32x32x8_F32", 1>;
+def MFMA_F16_32x32x8_F32  : I32EnumAttrCase<"MFMA_F16_32x32x8_F32", 1>;
+// TODO: Create separate WMMA ops for AMD and NVIDIA GPUs
 def WMMA_F16_16x16x16_F32 : I32EnumAttrCase<"WMMA_F16_16x16x16_F32", 2>;
+def WMMA_F16_16x16x16_F16 : I32EnumAttrCase<"WMMA_F16_16x16x16_F16", 3>;
```

Comment:
If this is supposed to be for cooperative matrix, we should define a separate attribute for it. This is not very clear from the descriptions below, but `IREEGPU_MMAIntrinsic` is effectively restricted to LLVMGPU. This is because lowering to `amgpu.wmma/mfma` (with an explicit layout) only makes sense for intrinsics used when targeting LLVM. We can make some of the interface methods on `IREEGPU_MmaInterfaceAttr` optional/FailureOr to allow opaque layouts. Then the KernelConfig logic will just look at shapes and doesn't have to care whether the mma kind is opaque or not.

Reply:
If it's possible to define external interface implementations on attributes, that sounds best and then we just define the interface on `#spirv.coop_matrix_props_khr`.

Reply:
The goal is to unify all the different GPU backends with one set of targets. So ideally we should have one `#iree_gpu.target` that works for both cuda/hip and vulkan. That's largely upheld now--if you look at how `getVulkanTargetDetails` and `getCUDATargetDetails`/`getHIPTargetDetails`, they go to the same underlying `getAMDTargetDtails`/`getNVIDIATargetDetails`.  This makes sure we have the same `wgp =` and `chip =` in `#iree_gpu.target`. It also conceptually makes sense--it's the same hardware no matter what apis we use to drive it. Though I saied before it's "largely upheld"--we have the `feature =` part of `#iree_gpu.target` doing separate things for separate underlying backends. E.g., `+sramecc,-xnack` for hip, `spirv:v1.3` for spirv, etc. That's more controlling what the llvm/spirv backend compiler in llvm/mlir. It's a bit annoying but I don't ever see a way to avoid that.

So I'd be hesitant to create separate mma attributes for separate gpu apis. we can create separate ones for separate vendors, which makes sense because they are different layouts. so that's what I meant we want to create separate ones for nvidia in the comment. even though they are opaque in the vulkan side; they certainly have a hardware layout. for amd side, we know exactly what wmma layout are. for nvidia side, wmma layout is not encoded in their ptx spec but you can figure it out like https://github.com/ahennequ/cuda-tensorcores-register-mapping. though if it's only for vulkan we might not need that info. so I agree with you that we may want to turn some interface methods on `IREEGPU_MmaInterfaceAttr` as optional.

Reply:
> we can create separate ones for separate vendors, which makes sense because they are different layouts. so that's what I meant we want to create separate ones for nvidia in the comment. even though they are opaque in the vulkan side; they certainly have a hardware layout. for amd side, we know exactly what wmma layout are.

+1

Reply:
Ok, I see your point. Agreed that excessive duplication is not wanted and somewhat defeats the purpose of this unification. I think it could still be worth adding a generic `opaque_mma` attribute so that plumbing through all possible intrinsics for a vendor doesn't have to blow up the enum set, assuming we don't want to jump straight to fully fledged non-opaque implementation, but makes sense to do that later as needed.

Reply:
Actually one other follow up, I see that the `vulkan-target-env` flag is being dropped, and only see a single WMMA type added here. Are the other (say WMMA_I8_16x16x16_I32) ones still targetable with IREE after this change?

Reply:
> I think it could still be worth adding a generic opaque_mma attribute so that plumbing through all possible intrinsics for a vendor doesn't have to blow up the enum set, assuming we don't want to jump straight to fully fledged non-opaque implementation, but makes sense to do that later as needed.

Yeah. I'm trying to cut at a reasonable place for a few things as follow ups and don't want to use one commit to change the whole world. We can do that as a next step. I put an item in https://github.com/iree-org/iree/issues/16341.

> Actually one other follow up, I see that the vulkan-target-env flag is being dropped, and only see a single WMMA type added here. Are the other (say WMMA_I8_16x16x16_I32) ones still targetable with IREE after this change?

It will fail support for i8 cases. We aren't supporting them anyway for Vulkan previously?

And I'm expecting we are gonna to add that on HIP side next for upcoming workloads so Vulkan picks it up "freely" anyway.